### DISCUSSION TITLE
Spherical joint using quaternion with external forces

### DISCUSSION THREAD

**Discussion by zoufaond** [sympy/sympy#26765] [Q&A] [answered]
Hello everyone,
let's say I have the equations of motion for single 3D pendulum (with spherical joint) in the following form:

$\textbf{M} \dot{u} = \textit{forcing}+\textit{fe}$

 where $\textit{fe}$ is a torque vector acting on the generalized coordinates (in my case these torques are calculated from muscle forces via moment arms to these generalized coordinates).

When I derive the same dynamical model using quaternion, the equations of motion are in the following form:

$\textbf{M} \dot{\omega} = \textit{forcing}$

I cannot just add my torques $\textit{fe}$ from muscles forces to the right side of the equation as these torques are acting on generalized coordinates, but the EOMs are in terms of angular velocities. My question is if there is some kind of way how to represent these torques so it acts as external moment applied to the rigid body. It is maybe more of a question regarding mechanics in general, I was just thinking that there is maybe a way to do the trick in sympy since I derive these equations in sympy. Thank you.

Here is the code for pendulum with quaternion.
```
import sympy as sp
import sympy.physics.mechanics as me

# symbols
t = sp.symbols('t')
q0, q1, q2, q3 = mech.dynamicsymbols('q0 q1 q2 q3')  # quaternion
u1, u2, u3 = mech.dynamicsymbols('u1 u2 u3')  # angular velocities
l, m, g, Ixx, Iyy, Izz = sp.symbols('l m g Ixx Iyy Izz')  # 

# quaternion
q = sp.Matrix([q0, q1, q2, q3])

# ref frame and body frame
N = mech.ReferenceFrame('N')
B = mech.ReferenceFrame('B')

# orient using quaternion
B.orient(N, 'Quaternion', q)

# set angular velocity
B.set_ang_vel(N, u1*B.x + u2*B.y + u3*B.z)

O = mech.Point('O')
O.set_vel(N, 0)

P = O.locatenew('P', -l * B.z)
P.v2pt_theory(O, N, B)

# body definiton
inertia_dyadic = mech.inertia(B, Ixx, Iyy, Izz)
pendulum = me.Body('pendulum', P, m, B, (inertia_dyadic))

gravity_force = (P, -m * g * N.z)


coordinates = [q0, q1, q2, q3]
speeds = [u1, u2, u3]

kinematic_differential_equations = [
    q0.diff(t) - 0.5 * (-u1*q1 - u2*q2 - u3*q3),
    q1.diff(t) - 0.5 * (u1*q0 + u3*q2 - u2*q3),
    q2.diff(t) - 0.5 * (u2*q0 - u3*q1 + u1*q3),
    q3.diff(t) - 0.5 * (u3*q0 + u2*q1 - u1*q2)
]

kane = mech.KanesMethod(N, q_ind=coordinates, u_ind=speeds, kd_eqs=kinematic_differential_equations)
fr, frstar = kane.kanes_equations([pendulum], [gravity_force,moment])
```

**Comment by moorepants** [comment, marked as answer]
Here is an example with quaternions and some forces applied:

https://opty.readthedocs.io/latest/examples/plot_drone.html#sphx-glr-examples-plot-drone-py

Maybe that helps.

In general with Kane's method you would write the forces however is convenient and they will be projected onto the generalized speed space in the transformation to generalized active forces. These means that they will not look like a simple force addition in the equations of motion.

**Reply by zoufaond** [reply]
Thanks, it is helpful as an example for building a model with quaternion. 
However these are forces in a ribidbody's frame. This way I could model the straight-line muscles as "point-to-point" forces acting in opposite direction on each point. I however derive the muscle forces using muscle excursion method, this way I calculate moment arm to a generalized coordinate (mostly these generalized coordinates are a sequence of Euler or Tait-Bryan angles). So I was wondering if there is some way to "combine " these torques acting on sequence of rotation with quaternion.

I tried to create 3D pendulum with 3 torques applied on a rigid body like this:
```
l,m,g,IBxx,IByy,IBzz, IBxy, IByz, IBzx, M1, M2, M3 = symbols('l m g IBxx IByy IBzz IBxy IByz IBzx M1 M2 M3')
q1, q2, q3 = dynamicsymbols('q1 q2 q3')
q1d, q2d, q3d = dynamicsymbols('q1 q2 q3', 1)
u1, u2, u3 = dynamicsymbols('u1 u2 u3')
u1d, u2d, u3d = dynamicsymbols('u1 u2 u3', 1)

masscenter = Point('masscenter')
frame = ReferenceFrame('frame1')
mass = Symbol('mass')
ceiling = Body('C')
body_inertia = inertia(frame, IBxx, IByy, IBzz, IBxy, IByz, IBzx)
bob = Body('U', masscenter, mass, frame, body_inertia)
joint = SphericalJoint('P1', ceiling, bob, child_point=l*bob.z, coordinates = [q1, q2, q3], speeds = [u1, u2, u3],rot_order='123', rot_type = 'body')
bob.apply_force(-ceiling.z*bob.mass*g)

# here are the torques
T1 = M1*frame.x
T2 = M2*frame.y
T3 = M3*frame.z

bob.apply_torque(T1)
bob.apply_torque(T2)
bob.apply_torque(T3)
```
In the equations of motion ($\textbf{M}\dot{\textit{u}} = \textit{fo}$) the torques are presented this way (for XYZ sequence):
```
forcing[0] = M1*cos(q2)*cos(q3) - M2*sin(q3)*cos(q2) + M3*sin(q2) + ...
forcing[1] = M1*sin(q3) + M2*cos(q3) + ...
forcing[2] = ... + M3 + ...
``` 
It can that be written that the contribution from M1,M2 and M3 to $\dot{\textit{u}}$ is via matrix 
```
[cos(q2)*cos(q3), -sin(q3)*cos(q2), sin(q2);
 sin(q3)        ,  cos(q3)         ,0;
 0              ,0             ,1]
```
So one solution (not really smart) would be to take the muscle moment forces acting on $\dot{\textit{u}}$, multiply it by inverse of this matrix to get the moment forces acting on a rigid body. It works, but it is a really sketchy solution and you don't really get rid of the gimbal joint as this matrix is not invertible when q2=pi. I thought you might now some trick in sympy to deal with this problem, or if you saw some examples regarding this problem. Thanks!

**Comment by moorepants** [comment]
Notice how I select the generalized speeds to be the body fixed angular velocity components in the drone example. This gives equations that look like:

```
⎢                      -c⋅vₓ(t) - g⋅m⋅(-2⋅q₀(t)⋅q₂(t) + 2⋅q₁(t)⋅q₃(t)) - m⋅(ω_y(t)⋅v_z(t) - ω_z(t)⋅v_y(t)) - m⋅──(vₓ(t))                       ⎥
⎢                                                                                                              dt                              ⎥
⎢                                                                                                                                              ⎥
⎢                                                                                                             d                                ⎥
⎢                      -c⋅v_y(t) - g⋅m⋅(2⋅q₀(t)⋅q₁(t) + 2⋅q₂(t)⋅q₃(t)) - m⋅(-ωₓ(t)⋅v_z(t) + ω_z(t)⋅vₓ(t)) - m⋅──(v_y(t))                       ⎥
⎢                                                                                                             dt                               ⎥
⎢                                                                                                                                              ⎥
⎢                     ⎛  2        2        2        2   ⎞                                       d                                              ⎥
⎢     -c⋅v_z(t) - g⋅m⋅⎝q₀ (t) - q₁ (t) - q₂ (t) + q₃ (t)⎠ - m⋅(ωₓ(t)⋅v_y(t) - ω_y(t)⋅vₓ(t)) - m⋅──(v_z(t)) + F₁(t) + F₂(t) + F₃(t) + F₄(t)     ⎥
⎢                                                                                               dt                                             ⎥
⎢                                                                                                                                              ⎥
⎢                                          ⎛ 2    2⎞ d                                                                                         ⎥
⎢              ⎛ 2    2⎞                 m⋅⎝d  + w ⎠⋅──(ωₓ(t))     ⎛ 2    2⎞                                                                   ⎥
⎢            m⋅⎝d  + l ⎠⋅ω_y(t)⋅ω_z(t)               dt          m⋅⎝l  + w ⎠⋅ω_y(t)⋅ω_z(t)   w⋅F₁(t)   w⋅F₂(t)   w⋅F₃(t)   w⋅F₄(t)             ⎥
⎢            ───────────────────────── - ───────────────────── - ───────────────────────── + ─────── + ─────── - ─────── - ───────             ⎥
⎢                        12                        12                        12                 2         2         2         2                ⎥
⎢                                                                                                                                              ⎥
⎢                                                        ⎛ 2    2⎞ d                                                                           ⎥
⎢                                                      m⋅⎝d  + l ⎠⋅──(ω_y(t))     ⎛ 2    2⎞                  ⎛ 2    2⎞                         ⎥
⎢              l⋅F₁(t)   l⋅F₂(t)   l⋅F₃(t)   l⋅F₄(t)               dt           m⋅⎝d  + w ⎠⋅ωₓ(t)⋅ω_z(t)   m⋅⎝l  + w ⎠⋅ωₓ(t)⋅ω_z(t)            ⎥
⎢            - ─────── + ─────── - ─────── + ─────── - ────────────────────── - ──────────────────────── + ────────────────────────            ⎥
⎢                 2         2         2         2                12                        12                         12                       ⎥
⎢                                                                                                                                              ⎥
⎢                                                                                          ⎛ 2    2⎞ d                                         ⎥
⎢                                    ⎛ 2    2⎞                  ⎛ 2    2⎞                m⋅⎝l  + w ⎠⋅──(ω_z(t))                                ⎥
⎢                                  m⋅⎝d  + l ⎠⋅ωₓ(t)⋅ω_y(t)   m⋅⎝d  + w ⎠⋅ωₓ(t)⋅ω_y(t)               dt                                        ⎥
⎢                                - ──────────────────────── + ──────────────────────── - ──────────────────────                                ⎥
⎢                                             12                         12                        12                                          ⎥
⎢                                                                                                                                              ⎥
⎢                                                      2        2        2        2                                                            ⎥
⎣                                                    q₀ (t) + q₁ (t) + q₂ (t) + q₃ (t) - 1                                                     ⎦

```

and you get the forces and moments in an easily interpret-able form.

**Reply by zoufaond** [reply]
Hi Jason,

I tried to do the basic 3D pendulum using the approach based on the drone example you posted and I have few questions.

So following the EOMs creation in the drone example I ended up modelling 3D pendulum like this:

```
A.orient_quaternion(N, (q0, q1, q2, q3))

N_w_A = A.ang_vel_in(N)

kinematical = sm.Matrix([
    u0 - q0.diff(t),
    wx - N_w_A.dot(A.x),
    wy - N_w_A.dot(A.y),
    wz - N_w_A.dot(A.z),
])

A.set_ang_vel(N, wx*A.x + wy*A.y + wz*A.z)

holonomic = sm.Matrix([q0**2 + q1**2 + q2**2 + q3**2 - 1])

kane = me.KanesMethod(
    N,
    (q1, q2, q3),
    (wx, wy, wz),
    kd_eqs=kinematical,
    q_dependent=(q0,),
    u_dependent=(u0,),
    configuration_constraints=holonomic,
    velocity_constraints=holonomic.diff(t),
)
```
Then I tried create the pendulum the way I posted:

```
# orient using quaternion
B.orient(N, 'Quaternion', q)

# set angular velocity
B.set_ang_vel(N, u1*B.x + u2*B.y + u3*B.z)

kinematic_differential_equations = [
    q0.diff(t) - 0.5 * (-u1*q1 - u2*q2 - u3*q3),
    q1.diff(t) - 0.5 * (u1*q0 + u3*q2 - u2*q3),
    q2.diff(t) - 0.5 * (u2*q0 - u3*q1 + u1*q3),
    q3.diff(t) - 0.5 * (u3*q0 + u2*q1 - u1*q2)
]

kane = me.KanesMethod(N, 
                      q_ind=coordinates, 
                      u_ind=speeds, 
                      kd_eqs=kinematic_differential_equations)
```

Both these methods yields to the same FR and FRSTAR (so I believe the latter is correct as well). The drone approach however has the advantage that the quaternion is automatically enforced to be unit (which is particulary handy with the direct collocation). I don't actually understand what is the real purpose of u0, and if this necesseraly needs to be there as now there are 8 states, whereas in my approach there are only 7 states. 
Also when I tried to to simulate the system (the one based on the drone approach) using `pydy.system`. There is a singularity when you set initial condition for quaternion as any of these [0,1,0,0], [0,0,1,0] and [0,0,0,1]. Could you help me understand why the singularity occurs ? Thank you.

**Reply by moorepants** [reply]
> The drone approach however has the advantage that the quaternion is automatically enforced to be unit

It is not automatically enforced to be zero, but it is ensured to be held zero when solving the optimal control problem.

If you forward simulate systems with quaternions you also have to enforce that each quaternion length is unity, otherwise your system will disorient over time.

> I don't actually understand what is the real purpose of u0, and if this necesseraly needs to be there as now there are 8 states, whereas in my approach there are only 7 states.

In general, in Kane's approach you introduce N generalized speeds where N is the # of coordinates . For a specific system, all kinematical differential equations may not be needed.  But here is the angular velocity of a body oriented by a quaternion:

```
In [21]: B.ang_vel_in(A).to_matrix(B)
Out[21]: 
Matrix([
[2*q0(t)*Derivative(q1(t), t) - 2*q1(t)*Derivative(q0(t), t) - 2*q2(t)*Derivative(q3(t), t) + 2*q3(t)*Derivative(q2(t), t)],
[2*q0(t)*Derivative(q2(t), t) + 2*q1(t)*Derivative(q3(t), t) - 2*q2(t)*Derivative(q0(t), t) - 2*q3(t)*Derivative(q1(t), t)],
[2*q0(t)*Derivative(q3(t), t) - 2*q1(t)*Derivative(q2(t), t) + 2*q2(t)*Derivative(q1(t), t) - 2*q3(t)*Derivative(q0(t), t)]])
```

You need kinematical differential equations for all four q's to ensure values for each q can be calculated. One choice is always u = q' which would give:

```
In [22]: B.ang_vel_in(A).to_matrix(B).xreplace({qi.diff(): ui for qi, ui in zip((q0, q1, q2, q3), (u0, u1, u2, u3))})
Out[22]: 
Matrix([
[2*q0(t)*u1(t) - 2*q1(t)*u0(t) - 2*q2(t)*u3(t) + 2*q3(t)*u2(t)],
[2*q0(t)*u2(t) + 2*q1(t)*u3(t) - 2*q2(t)*u0(t) - 2*q3(t)*u1(t)],
[2*q0(t)*u3(t) - 2*q1(t)*u2(t) + 2*q2(t)*u1(t) - 2*q3(t)*u0(t)]])
```

If I use your definitions you get:

```
In [23]: kinematic_differential_equations = [
    ...:     q0.diff(t) - 0.5 * (-u1*q1 - u2*q2 - u3*q3),
    ...:     q1.diff(t) - 0.5 * (u1*q0 + u3*q2 - u2*q3),
    ...:     q2.diff(t) - 0.5 * (u2*q0 - u3*q1 + u1*q3),
    ...:     q3.diff(t) - 0.5 * (u3*q0 + u2*q1 - u1*q2)
    ...: ]

In [24]: qd_repl = sm.solve(kinematic_differential_equations, (q0.diff(), q1.diff(), q2.diff(), q3.diff()))

In [27]: B.ang_vel_in(A).xreplace(qd_repl).simplify().to_matrix(B)
Out[27]: 
Matrix([
[(q0(t)**2 + q1(t)**2 + q2(t)**2 + q3(t)**2)*u1(t)],
[(q0(t)**2 + q1(t)**2 + q2(t)**2 + q3(t)**2)*u2(t)],
[(q0(t)**2 + q1(t)**2 + q2(t)**2 + q3(t)**2)*u3(t)]])
```

which eliminates the extra generalized speed.

> There is a singularity when you set initial condition for quaternion as any of these [0,1,0,0], [0,0,1,0] and [0,0,0,1]. Could you help me understand why the singularity occurs ? Thank you.

This is common when using Kane's method due to solving any motion constraints symbolically. You can see a divide by zero here if q0=0:

```
In [28]: holonomic = sm.Matrix([q0**2 + q1**2 + q2**2 + q3**2 - 1])

In [29]: holonomic.diff(t)
Out[29]: Matrix([[2*q0(t)*Derivative(q0(t), t) + 2*q1(t)*Derivative(q1(t), t) + 2*q2(t)*Derivative(q2(t), t) + 2*q3(t)*Derivative(q3(t), t)]])

In [31]: sm.solve(holonomic.diff(t), q0.diff())
Out[31]: {Derivative(q0(t), t): -q1(t)*Derivative(q1(t), t)/q0(t) - q2(t)*Derivative(q2(t), t)/q0(t) - q3(t)*Derivative(q3(t), t)/q0(t)}
```

**Reply by zoufaond** [reply]
Both of the solutions yields to the same result though, also the solution of mine does not have the singularities. Does it work only because the system is easy and unconstrained, and if there were some constraints, then you would need to strictly use your approach ?

I am just trying to understand why my solution works as well. I interpret it this way: the rigid body has 3 coordinates for rotation and for angular velocity, which leads to 3 speeds. The quaternion has 4 degrees of freedom but it has to be unit which removes 1 dof.

**Reply by moorepants** [reply]
I don't think you need to use my approach with u0 = q0.diff(t). But you will have to ensure `(q0(t)**2 + q1(t)**2 + q2(t)**2 + q3(t)**2) = 1`.

**Reply by zoufaond** [reply]
Ok, thanks for your answers

**Comment by oscarbenjamin** [comment]
> If you forward simulate systems with quaternions you also have to enforce that each quaternion length is unity, otherwise your system will disorient over time.

Just an aside here but Gauss-Legendre methods such as implicit midpoint eliminate the truncation error for preserving quadratic invariants like the unit norm constraint for a quaternion. That means that the norm will be preserved up to small rounding errors if implicit midpoint is used.

**Reply by zoufaond** [reply]
Thanks for the useful note. I've already read somewhere that renormalizing the quaternion each step (when using explicit ode solvers) is not the best approach.

**Reply by moorepants** [reply]
It's not the best approach for any algebraic constraints, including the quaternion length constraint.
